Fix MustNotThrow annotations and add invariant to avoid exception handling in __CPROVER_initialize

owen-mc-diffblue opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add MustNotThrow annotations to Class.java in tests

These are cut-down versions of the model for java.lang.Class. The
methods that are kept are now marked MustNotThrow in the model, so
that exception-handling code isn't generated in __CPROVER_initialize.
We should therefore update the cut-down versions.

smowton commented on jbmc/src/java_bytecode/remove_exceptions.cpp:
```diff
@@ -587,6 +589,14 @@ void remove_exceptionst::instrument_exceptions(
+  // @inflight_exception is initialized; (2) symex can analyze
+  // INITIALIZE_FUNCTION much faster if it doesn't contain any branching.
+  INVARIANT(function_identifier != INITIALIZE_FUNCTION || !did_something,
+            INITIALIZE_FUNCTION "should not contain any exception handling branches");
```
`" should`

smowton reviewed (approved):
Once CI is passing, of course

tautschnig commented on jbmc/regression/jbmc/class-fields/java/lang/Class.java:
```diff
@@ -4,6 +4,7 @@
 
   public Integer field;
 
+  @org.cprover.MustNotThrow
```
Wouldn't you also need to include updated .class files in the commit?

tautschnig commented on jbmc/src/java_bytecode/remove_exceptions.cpp:
```diff
@@ -587,6 +589,14 @@ void remove_exceptionst::instrument_exceptions(
+  // references @inflight_exception might be generated before
+  // @inflight_exception is initialized; (2) symex can analyze
+  // INITIALIZE_FUNCTION much faster if it doesn't contain any branching.
+  INVARIANT(function_identifier != INITIALIZE_FUNCTION || !did_something,
```
Given that seemingly you had to modify the source code of a test to make this `INVARIANT` pass I am wondering whether this really is an invariant?

owen-mc-diffblue replied:
What macro would capture my intent better? Quite a lot of work went into making sure that branches for exception-handling weren't created in `__CPROVER_initialize`. We want to make sure it doesn't creep back in by having an assertion which fails when it does.

tautschnig replied:
If user input can violate this constraint (which I thought was possible, given you had to change a test), then it should be an exception (for example ` unsupported_operation_exceptiont`).

owen-mc-diffblue replied:
The "user input" in those tests is a cut-down model for java.lang.Class, implementing special CPROVER functions like `cproverNondetInitialize` or `cproverInitializeClassLiteral`. I don't think we expect users to be providing anything like that.

tautschnig replied:
Oh, I see, I should probably just remain quiet about any Java-related code :-)  Thanks for clarifying this, the invariant does seem to make a lot of sense in that case.

owen-mc-diffblue commented:
I should have marked this `do not review` - I will fix it up when I have finished #4288 .

Commit:
Mark param as [out] in doxygen

Commit:
Ensure no exception handling in __CPROVER_initialize

__CPROVER_initialize should not contain any exception handling branches
for two reasons: (1) with symex-driven lazy loading it means that code
that references @inflight_exception might be generated before
@inflight_exception is initialized; (2) symex can analyze
INITIALIZE_FUNCTION much faster if it doesn't contain any branching.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5aaaf25a58a6bc9e15d9a56d35365a7d323fee59).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105834354

owen-mc-diffblue commented:
This is now ready for review

smowton reviewed (approved):
I'd switch `_GOTOS` for `_MAY_THROW` or similar wording, since it's throwing rather than gotos that are actually the problem. Otherwise lgtm.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 58ded6248a60cc865b6ffc56f7e4546cbd92cd5c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105890480
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
